Lushu/src/main/kotlin/lushu/ContextGrammar/Grammar/Rules.kt

```diff
@@ -0,0 +1,210 @@
+        var i: Int = 0
+        while(i < line.size){
+            current = matcher(current, line[i])
+
+            if(current == null){
+                i += 1
+            }
+            else{
+                if(current.isSensitive()){
+                    auxiliarLine[i] = "*".repeat(auxiliarLine[i].length)
+                }
+
+                i += match(current, (i+1))
+            }
+
+            current = root
+        }
+        return line
+    }
+    fun addRules(contextRule: MutableList<String>){
+        var current: Node? = root
+
+        var sensitive: Boolean = false
+        var plusCase: Boolean= false
+        var mergeable: Boolean = false
+
+        var nextSensitive: Boolean = false
+        var nextPlusCase: Boolean= false
+        var nextMergeable: Boolean = false
+
+        for(i in 0 until contextRule.size){
+            sensitive = nextSensitive
+            mergeable = nextMergeable
+            plusCase = nextPlusCase
+
+            if(contextRule[i].startsWith("<c>")){
+                contextRule.add(i, contextRule[i].substring(sizeInitialFlag, contextRule[i].length))
+                contextRule.removeAt(i+1)
+            }
+            if(contextRule[i].startsWith("<s>")){
+                contextRule.add(i, contextRule[i].substring(sizeInitialFlag, contextRule[i].length))
+                contextRule.removeAt(i+1)
+                sensitive = true
+                nextSensitive = true
+            }
+            if(contextRule[i].startsWith("<+>")){
+                contextRule.add(i, contextRule[i].substring(sizeInitialFlag, contextRule[i].length))
+                contextRule.removeAt(i+1)
+                plusCase = true
+                nextPlusCase = true
+            }
+            if(contextRule[i].startsWith("<t>")){
+                contextRule.add(i, contextRule[i].substring(sizeInitialFlag, contextRule[i].length))
+                contextRule.removeAt(i+1)
+                mergeable = true
+                nextMergeable = true
+            }
+            if(contextRule[i].endsWith("</c>")){
+                contextRule.add(i, contextRule[i].substring(0, contextRule[i].length - sizeFinalFlag))
+                contextRule.removeAt(i+1)
+            }
+            if(contextRule[i].endsWith("</s>")){
+                contextRule.add(i, contextRule[i].substring(0, contextRule[i].length - sizeFinalFlag))
+                contextRule.removeAt(i+1)
+                nextSensitive = false
+            }
+            if(contextRule[i].endsWith("</+>")){
+                contextRule.add(i, contextRule[i].substring(0, contextRule[i].length - sizeFinalFlag))
+                contextRule.removeAt(i+1)
+                nextPlusCase = false
+            }
+            if(contextRule[i].endsWith("</t>")){
+                contextRule.add(i, contextRule[i].substring(0, contextRule[i].length - sizeFinalFlag))
+                contextRule.removeAt(i+1)
+                nextMergeable = false
+            }
```

Comment:
We can probably simplify this portion of the code using a composite or factory design pattern.